### PR TITLE
Performance improvements when query includes sys.indexes.is_primary_key predicate 

### PR DESCRIPTION
**Pull request by tanyagupta17**
### Description
Previously for a query we were observing a performance degradation when it included a join between sys.indexes,  sys.index_columns, sys.columns, sys.tables, and sys.schemas views with sys.indexes.is_primary_key = 1 predicate in where clause. We identified that there was the issue with a case statement due to which there was an incorrect row estimation so we have implemented a direct cast by defining new IMPLICIT cast without function from boolean to sys.bit.
Another issue was the use of redundant has_schema_privilege in sys views and functions, as we explicitly grant usage to public beforehand, so we have removed the redundant `has_schema_privilege` check. 

Signed-off-by: Tanya Gupta [tanyagp@amazon.com](mailto:tanyagp@amazon.com)

### Issues Resolved

BABEL-5215

Performance test:
Overall execution time(in sec) for 102 schemas, 2319 tables, 2535 key_constraints, 5885 indexes and 2034 types

|Query|Before|After|
|---|---|---|
|Targeted Query|51|1.7|
|SSMS query to enumerate tables|1140|26|

### Test Scenarios Covered ###
* **Use case based -** Yes 

* **Boundary conditions -** Yes

* **Arbitrary inputs -** Yes

* **Negative test cases -** Yes

* **Minor version upgrade tests -** Yes

* **Major version upgrade tests -** Yes

* **Performance tests -** Yes

* **Tooling impact -** N/A

* **Client tests -** N/A

### Check List
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by sumitj824 — COMMENTED** [review]
Please follow the guidelines for good PR title and description.

File: contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.3.0--4.4.0.sql [hunk 1]
```diff
@@ -63,6 +63,9 @@ WITH FUNCTION sys.varchar2time(sys.VARCHAR, INT4) AS IMPLICIT;
 
 CALL sys.babelfish_drop_deprecated_object('function', 'sys', 'varchar2time_deprecated_4_4_0');
 
+CREATE CAST (bool AS sys.BIT)
+WITHOUT FUNCTION AS IMPLICIT;
+
```
Comment (sumitj824) [inline contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.3.0--4.4.0.sql]: As these performance improvements will be part of 3_X_DEV as well, upgrading from 3.8.0 to 4.4.0 will cause failures due to the existing cast. Can we add a check to create the cast only if it doesn't exist?

File: contrib/babelfishpg_tsql/sql/sys_views.sql [hunk 2]
```diff
@@ -72,7 +72,6 @@ left join tt_internal tt on t.oid = tt.typrelid
 where tt.typrelid is null
 and (t.relkind = 'r' or t.relkind = 'p')
 and t.relispartition = false
-and has_schema_privilege(t.relnamespace, 'USAGE')
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/sql/sys_views.sql:75]: There are still few occurences like sys.sp_tables_view, sys.sp_special_columns_view, etc. where you missed to remove the `has_schema_privilege(USAGE)` redundant check. Please make sure we remove it from everywhere.

**Review by sumitj824 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/sql/sys_views.sql (hunk 2 above)
Comment (tanyagupta17) [reply contrib/babelfishpg_tsql/sql/sys_views.sql:75]: Removed.

File: contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.3.0--4.4.0.sql (hunk 1 above)
Comment (tanyagupta17) [reply contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.3.0--4.4.0.sql]: Added

File: contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.3.0--4.4.0.sql [hunk 3]
```diff
@@ -63,6 +63,24 @@ WITH FUNCTION sys.varchar2time(sys.VARCHAR, INT4) AS IMPLICIT;
...
+    sys_oid Oid;
+    pg_catalog_oid Oid;
+    bool_oid Oid;
+    bit_oid Oid;
+BEGIN
+  sys_oid := (SELECT oid FROM pg_namespace WHERE pg_namespace.nspname ='sys');
+  pg_catalog_oid := (SELECT oid FROM pg_namespace WHERE pg_namespace.nspname ='pg_catalog');
+  bool_oid := (SELECT oid FROM pg_type WHERE pg_type.typname ='bool' AND pg_type.typnamespace = pg_catalog_oid);
+  bit_oid := (SELECT oid FROM pg_type WHERE pg_type.typname ='bit' AND pg_type.typnamespace = sys_oid);
+  IF (SELECT COUNT(*) FROM pg_cast WHERE pg_cast.castsource = bool_oid AND pg_cast.casttarget = bit_oid) = 0 THEN
```
Comment (sumitj824) [inline contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--4.3.0--4.4.0.sql:78]: Should we add restriction on `castcontext` and `castmethod` also?

**Review by sumitj824 — COMMENTED** [review]
Please add a table in description for achieved performance improvements.

**Review by sumitj824 — APPROVED** [review]
LGTM

File: test/JDBC/input/views/sys-syscolumns.sql [hunk 4]
```diff
@@ -1,4 +1,4 @@
--- sla 16000
+-- sla 20000
```
Comment (KushaalShroff) [inline test/JDBC/input/views/sys-syscolumns.sql]: Where are we testing the privileges that display that schema permissions is really redundant?

**Review by KushaalShroff — CHANGES_REQUESTED** [review]

File: test/JDBC/input/views/sys-syscolumns.sql (hunk 4 above)
Comment (tanyagupta17) [reply test/JDBC/input/views/sys-syscolumns.sql]: Added test cases to verify for sys.views and sys.all_objects in sys-views-vu-* files 

I have tested it manually in mammoth as well, with and without my changes. For a user created view and a particular user login we should now see the created view unless given the explicit grant select on schema to the user. 
The same behavior is observed after our changes as well

Comment (sumitj824) [reply test/JDBC/input/views/sys-syscolumns.sql]: @tanyagupta17 I think the tests you have added are already presents in multiple files. Here, we want to validate that `has_schema_privilege(USAGE)`  checks will always return true for all TSQL schemas and TSQL roles. So, you can invoke this function from PSQL endpoint in tests to validate the same.

Comment (tanyagupta17) [reply test/JDBC/input/views/sys-syscolumns.sql]: Done. Added the test case to check that all tsql users have USAGE privilege on tsql schemas

**Review by sumitj824 — APPROVED** [review]
LGTM with test changes.

**Review by KushaalShroff — APPROVED** [review]

**Pull request merged**